kusto/kcsb.go

```diff
@@ -0,0 +1,354 @@
+	"msi_auth":               msiAuth,
+	"managedserviceidentity": managedServiceIdentity, "managed service identity": managedServiceIdentity,
+	"interactive login": interactiveLogin, "interactivelogin": interactiveLogin,
+	"az cli": azCli, "azcli": azCli,
+	"domain hint": domainHint, "domainhint": domainHint,
+}
+
+func requireNonEmpty(key string, value string) {
+	if isEmpty(value) {
+		panic(fmt.Sprintf("Error: %s cannot be null", key))
+	}
+}
+
+func assignValue(kcsb *connectionStringBuilder, rawKey string, value string) {
+	rawKey = strings.ToLower(strings.Trim(rawKey, " "))
+	parsedKey := csMapping[rawKey]
+	switch parsedKey {
+	case dataSource:
+		kcsb.dataSource = value
+	case aadUserId:
+		kcsb.aadUserID = value
+	case password:
+		kcsb.password = value
+	case applicationClientId:
+		kcsb.applicationClientId = value
+	case applicationKey:
+		kcsb.applicationKey = value
+	case applicationCertificate:
+		kcsb.applicationCertificate = value
+	case applicationCertificateThumbprint:
+		kcsb.applicationCertificateThumbprint = value
+	case sendCertificateChain:
+		bval, _ := strconv.ParseBool(value)
+		kcsb.sendCertificateChain = bval
+	case authorityId:
+		kcsb.authorityId = value
+	case applicationToken:
+		kcsb.applicationToken = value
+	case userToken:
+		kcsb.userToken = value
+	case msiAuth:
+		bval, _ := strconv.ParseBool(value)
+		kcsb.msiAuthentication = bval
+	case managedServiceIdentity:
+		kcsb.managedServiceIdentity = value
+	case azCli:
+		bval, _ := strconv.ParseBool(value)
+		kcsb.azcli = bval
+	case interactiveLogin:
+		bval, _ := strconv.ParseBool(value)
+		kcsb.interactiveLogin = bval
+	case domainHint:
+		kcsb.redirectURL = value
+	}
+}
+
+/*
+   Creates new KustoConnectionStringBuilder.
+   Params takes kusto connection string connStr: string.  Kusto connection string should be of the format:
+   https://<clusterName>.kusto.windows.net;AAD User ID="user@microsoft.com";Password=P@ssWord
+   For more information please look at:
+   https://docs.microsoft.com/azure/data-explorer/kusto/api/connection-strings/kusto
+*/
+func GetConnectionStringBuilder(connStr string) connectionStringBuilder {
+	kcsb := connectionStringBuilder{}
+	if isEmpty(connStr) {
+		panic("Error : Connection string cannot be empty")
+	}
+
+	if !strings.Contains(strings.Split(connStr, ";")[0], "=") {
+		connStr = "Data Source=" + connStr
+	}
+
+	for _, kvp := range strings.Split(connStr, ";") {
+		kvparr := strings.Split(kvp, "=")
+		val := strings.Trim(kvparr[1], " ")
+		if isEmpty(val) {
+			continue
+		}
+		assignValue(&kcsb, kvparr[0], val)
+
+	}
+
+	return kcsb
+}
+
+//Creates a KustoConnection string builder that will authenticate with AAD user name and password.
+func (kcsb connectionStringBuilder) WithAadUserPassAuth(uname string, pswrd string, authorityID string) connectionStringBuilder {
+	requireNonEmpty(dataSource, kcsb.dataSource)
+	requireNonEmpty(aadUserId, uname)
+	requireNonEmpty(password, pswrd)
+	kcsb.aadUserID = uname
```

Comment:
C# SDK rightfully clears all auth properties before assigning anything in WithXXX.
This means calling With resets previous auth options whether they came from the string or from a previous call to With